lib/compat/wordpress-6.2/class-gutenberg-rest-themes-controller-6-2.php

```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * REST API: Gutenberg_REST_Themes_Controller class
+ *
+ * @package    Gutenberg
+ * @subpackage REST_API
+ */
+
+/**
+ * Core class used to manage themes via the REST API.
+ *
+ * @since 5.0.0
+ *
+ * @see WP_REST_Controller
+ */
+class Gutenberg_REST_Themes_Controller_6_2 extends WP_REST_Themes_Controller {
+	/**
+	 * Prepares links for the request.
+	 *
+	 * @since 5.7.0
+	 *
+	 * @param WP_Theme $theme Theme data.
+	 * @return array Links for the given block type.
+	 */
+	protected function prepare_links( $theme ) {
```

Comment:
This method is equal to the upstream at https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php#L358 but using the gutenberg classes.

Reply:
Yep, because the regular WP_Theme_JSON_Resolver fetches wp_global_styles posts in descending order. The new functionality assumes that the "current user global styles" are always the first wp_global_styles post that is made (per theme).

Reply:
So. I've been educating myself about how this currently works and the changes here. Also checked [the issue](https://github.com/WordPress/gutenberg/issues/45371) and [original PR](https://github.com/WordPress/gutenberg/pull/46952) for context. Let me know if I should check any other things.

Would this be a fair representation of how this works?

- The way this PR adds support for user style variations is by storing them as new `wp_global_styles` CPTs.
- Before this PR, there was only a single `wp_global_styles` CPT per theme. We took those user styles to merge them with the others and generate the global stylesheet.
- Now, there can be multiple `wp_global_styles` CPT: as many as user style variations + user styles if no style variation is selected.

Once we store everything under the same CPT, **how do tell apart the "user styles" from the "user style variations"?**

This PR proposes the "user styles" to be the older CPT. Can this be problematic? For example, if the post date changes in uncontrolled situations (exporting/importing the data to a different site?). I am not certain which specific example would break this, though I'd be more comfortable if we can tell apart "user styles" CPT by something explicit, that cannot change based on external factors.

Are there alternatives to using `post_date`? For example, what would you think of using the `post_status` field? Perhaps the "user styles" can have `post_status=publish` while the "user style variations" use the `post_status=draft`.

Reply:
I'll be using the terminology established in the main comment of this issue (I made some edits). CUGS is different from custom styles by being the oldest and by having an `associated_style_id` meta key. I think being the oldest is sufficient for all use cases. Even if export / import happens, the post date should be maintained. And if the export / import is done in a way that post_date is not maintained, who is to say that post_status will be?

We could make this differentiation more strongly enforced by also querying for the existence of 'associated_style_id meta key whenever we're querying for CUGS, however adding a meta query makes the SQL slower so I didn't do this because I don't see a reasonable scenario where having the oldest post_date would fail.

Reply:
I slept on the use of `post_date` as an identifier for the user styles, and I'm still unconvinced it's safe. By depending on an external factor (the clock in a computer) the code is not deterministic: if the clock of any computer that creates global styles goes out of sync, this code will certainly introduce errors. At the scale of WordPress, that's bound to happen at some point, somewhere. [This](https://g13g.blog/2017/03/16/my-favorite-bug/) is a "fun" bug that happened because of clocks! That's what I want to avoid.

I'm not sure about what's the best alternative, though, so happy to explore any. This is something we need to figure it out _before_ deploying this. It's not something we can fix _after_. Once there's user data in the database, it's a lot harder to fix (if at all possible).

Reply:
Alrighty. How about we go for the approach to add `cugs_id_{$stylesheet}` to `wp_options`?

Reply:
re: terminology. I'd appreciate if we can stick to the existing nomenclature:

- _user styles_ for the data that is merged with core, blocks, and theme styles to generate the global stylesheet
- _user style variations_: for the different set of styles already preconfigured

The WordPress styles sub-system is already a quite loaded domain and a there has been a few people working on it for a while. Changing/Updating established terminology is bound to be confusing.

Reply:
I updated the main post comment with these terminology definitions so things are clearer, I will use that terminology going forward but I won't edit all of my other comments.

In that case, the meta key would be called `user_styles_id_{$stylesheet}` in the database. Should we do that?

Reply:
> Alrighty. How about we go for the approach to add cugs_id_{$stylesheet} to wp_options?

Do we have alternatives that involve extracting that information from the data we already have instead of querying/linking options? I mean, I don't think it'd be a big deal, though it'd be nice/simpler if we don't have to.

In light of this [other conversation](https://github.com/WordPress/gutenberg/pull/47075#discussion_r1071975107), do you think it'd be enough to use the `post_name`? 

- The user styles CPT would be stored as `wp-global-styles-${stylesheet}`.
- The style variations CPTs would be stored as `wp-global-styles-${stylesheet}-variation-${variation-identifier}`.

Reply:
@oandregal I made this change. Now the user styles CPT is stored as `wp-global-styles-${stylesheet}` whereas custom style variations get a different name. Also, when user styles CPT is queried, it checks the post name, so now it no longer depends only on the post date but also on the post name.